### PR TITLE
Write Frame OBUs instead of a pair of Frame Header and Tile Group OBUs

### PR DESCRIPTION
**Pull request by jamrial**
For non show_existing_frame frames, and seeing rav1e currently doens't offer
the option to write more than one Tile Group per frame, it's more
efficient to use Frame OBUs instead of a pair of Frame Header and Tile Group
OBUs.

This saves a few bytes per frame by removing the overhead from the superfluous
OBU header.

### CODE REVIEW

**Comment by jamrial** [comment]
Could someone take a look at this?

File: src/encoder.rs [hunk 1]
```diff
@@ -3449,9 +3419,37 @@ pub fn encode_show_existing_frame<T: Pixel>(
   fi: &FrameInvariants<T>, fs: &mut FrameState<T>, inter_cfg: &InterConfig,
 ) -> Vec<u8> {
   debug_assert!(fi.show_existing_frame);
+  let obu_extension = 0 as u32;
```
Comment (tmatth) [inline src/encoder.rs:3422]: Any reason for not using `0u32` ?

File: src/encoder.rs [hunk 2]
```diff
@@ -3484,6 +3482,8 @@ pub fn encode_frame<T: Pixel>(
 ) -> Vec<u8> {
   debug_assert!(!fi.show_existing_frame);
   debug_assert!(!fi.invalid);
+  let obu_extension = 0 as u32;
```
Comment (tmatth) [inline src/encoder.rs:3485]: idem

**Review by tmatth — COMMENTED** [review]
LGTM, just some minor style questions, but someone more familiar with OBUs should take a look (@rzumer ?)

File: src/encoder.rs (hunk 1 above)
Comment (jamrial) [reply src/encoder.rs:3422]: Not familiar enough with Rust, so i wasn't aware that's an option

Comment (lu-zero) [reply src/encoder.rs:3422]: It can be `0`, the type should inferred w/out external help.

File: src/encoder.rs [hunk 3]
```diff
@@ -3449,9 +3419,37 @@ pub fn encode_show_existing_frame<T: Pixel>(
...
   debug_assert!(fi.show_existing_frame);
+  let obu_extension = 0;
+
   let mut packet = Vec::new();
 
-  write_obus(&mut packet, fi, fs, inter_cfg).unwrap();
+  if fi.frame_type == FrameType::KEY {
+    write_key_frame_obus(&mut packet, fi, obu_extension).unwrap();
+  }
+  let mut buf1 = Vec::new();
```
Comment (rzumer) [inline src/encoder.rs:3429]: Why are those two defined and then later redefined in unconditional inner blocks?

File: src/encoder.rs [hunk 4]
```diff
@@ -3449,9 +3419,37 @@ pub fn encode_show_existing_frame<T: Pixel>(
   fi: &FrameInvariants<T>, fs: &mut FrameState<T>, inter_cfg: &InterConfig,
 ) -> Vec<u8> {
   debug_assert!(fi.show_existing_frame);
+  let obu_extension = 0;
```
Comment (rzumer) [inline src/encoder.rs:3422]: What is this constant's purpose?

File: src/header.rs [hunk 5]
```diff
@@ -843,7 +843,6 @@ impl<W: io::Write> UncompressedHeader for BitWriter<W, BigEndian> {
     if fi.large_scale_tile {
       unimplemented!();
     }
-    self.write_bit(true)?; // trailing bit
```
Comment (rzumer) [inline src/header.rs:846]: Why is this removed?

File: src/encoder.rs [hunk 6]
```diff
@@ -3492,22 +3492,33 @@ pub fn encode_frame<T: Pixel>(
...
-    bw1.write_obu_header(ObuType::OBU_TILE_GROUP, 0).unwrap();
+    bw1.write_obu_header(ObuType::OBU_FRAME, obu_extension).unwrap();
   }
   packet.write_all(&buf1).unwrap();
   buf1.clear();
 
   {
     let mut bw1 = BitWriter::endian(&mut buf1, BigEndian);
-    bw1.write_uleb128(tile_group.len() as u64).unwrap();
+    bw1.write_uleb128((buf2.len() + tile_group.len()) as u64).unwrap();
```
Comment (rzumer) [inline src/encoder.rs:3514]: I can't tell what's going on here. It is writing data from `buf1` but for the length of `buf2`. Why?

File: src/encoder.rs [hunk 7]
```diff
@@ -3492,22 +3492,33 @@ pub fn encode_frame<T: Pixel>(
   }
   let tile_group = encode_tile_group(fi, fs, inter_cfg);
 
-  write_obus(&mut packet, fi, fs, inter_cfg).unwrap();
+  if fi.frame_type == FrameType::KEY {
+    write_key_frame_obus(&mut packet, fi, obu_extension).unwrap();
+  }
```
Comment (rzumer) [inline src/encoder.rs:3497]: Would like another line break after this to separate steps and make reading this easier.

File: src/encoder.rs [hunk 8]
```diff
@@ -3449,9 +3419,37 @@ pub fn encode_show_existing_frame<T: Pixel>(
...
+
+  {
+    let mut bw1 = BitWriter::endian(&mut buf1, BigEndian);
+    bw1.write_obu_header(ObuType::OBU_FRAME_HEADER, obu_extension).unwrap();
+  }
+  packet.write_all(&buf1).unwrap();
+  buf1.clear();
+
+  {
+    let mut bw1 = BitWriter::endian(&mut buf1, BigEndian);
```
Comment (rzumer) [inline src/encoder.rs:3444]: Here there are two consecutive uses of `bw1`, they can probably be merged.

File: src/encoder.rs (hunk 3 above)
Comment (rzumer) [inline src/encoder.rs:3429]: Not really in scope for this PR since it matches the previous state of the code, but I'd like some clarity about why we have two buffers and writers, through renaming the variables or comments. It's not clear at a glance why this code is written the way it is.

File: src/encoder.rs (hunk 4 above)
Comment (jamrial) [reply src/encoder.rs:3422]: I moved it around when i refactored these functions. It looked superfluous to me as well, but i figured leaving it alone was the better idea, in case verbosity was preferred.

File: src/header.rs (hunk 5 above)
Comment (jamrial) [reply src/header.rs:846]: Because it's no longer needed. That bit is the trailing bit at the end of a Frame Header to signal the end of the entire OBU, and we don't write those anymore. Frame OBUs are composed by a Frame Header followed immediately by a Tile Group merged into a single OBU.

See section 5.10 in the spec.

File: src/encoder.rs (hunk 4 above)
Comment (rzumer) [reply src/encoder.rs:3422]: I see, I missed its definition in the previous version of the code. I don't have a strong opinion on how to handle it.

File: src/encoder.rs (hunk 8 above)
Comment (jamrial) [reply src/encoder.rs:3444]: It's the same as before this patch, and i can't seem to find a way to merge them.

Comment (rzumer) [reply src/encoder.rs:3445]: Okay, not important for this patch then, I will look into it on my own.

File: src/encoder.rs (hunk 7 above)
Comment (jamrial) [reply src/encoder.rs:3498]: Done

**Comment by rzumer** [comment]
AWCY run results: http://35.223.44.203/?job=master%402020-11-18T21%3A18%3A35.622Z&job=obu-optimization%402020-11-18T21%3A29%3A25.667Z

**Comment by jamrial** [comment]
I was not expecting changes in encoding metrics. This simply saves a few overhead header bytes per frame.

**Comment by rzumer** [comment]
I can't say why the metrics might have changed, and the analyzer is not hooked up to this AWCY instance, but I started a new run prior to the last commit on master which is not merged into your branch, in case it caused unexpected coding changes: http://35.223.44.203/?job=master-minus-one%402020-11-18T22%3A27%3A40.206Z

**Review by rzumer — APPROVED** [review]

**Comment by rzumer** [comment]
Just to follow up on the initial results, it turned out the "master" reference did not point to the origin repository as expected.
